experimental/torch_xla2/torch_xla2/tensor.py

```diff
@@ -281,3 +239,155 @@ def debug_accuracy(func, args, kwargs, current_output):
       pdb.set_trace()
 
   return True
+
+
+class XLAFunctionMode(torch.overrides.TorchFunctionMode):
+  """Context manager that dispatches torch function calls to JAX."""
+
+  def __init__(self, env):
+     self.env = env
+
+  def __torch_function__(self,
+                         func,
+                         types,
+                         args=(),
+                         kwargs=None) -> torch.Tensor:
+    try:
+      return self.env.dispatch(func, types, args, kwargs)
+    except OperatorNotFound:
+      return func(*args, **(kwargs or {}))
+
+
+class XLADispatchMode(torch_dispatch.TorchDispatchMode):
+
+  def __init__(self, env):
+    self.env = env
+
+  def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+    if isinstance(func, torch._ops.OpOverloadPacket):
+      with self:
+        return func(*args, **kwargs)
+    if func.namespace != 'aten':
+      return func(*args, **kwargs)
+    return self.env.dispatch(func, types, args, kwargs)
+
+def _name_of_func(func):
+  if hasattr(func, 'name'):
+    return func.name()
+  return func.__name__
+
+
+class Environment:
```

Comment:
Can the environment inherit `TorchDispatchMode` and `TorchFunctionMode` and implement `__torch_function__` and `__torch_dispatch__` directly? That would change the semantic to `with torch_xla2.default_env():`

Or do we ever expect to use `XLAFunctionMode` and `XLADispatchMode` independently of `Environment`?

Reply:
Making Environment a context manager itself makes sense. However, multiple inheritance won't work as only one mode will activate. on calling of `__enter__` with `with`, Python's method resolution will find the first parent with that method and skip the second parent. I added `__enter__` and `__exit__` to Environment class instead.